File: 24-01-2025/java/Valid Parentheses.java
//https://leetcode.com/problems/valid-parentheses/description/

//code
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();

        for (char c : s.toCharArray()) {
            if (c == '(' || c == '{' || c == '[') {
                stack.push(c);
            }
           else if (stack.isEmpty() ||
                    (c == ')' && stack.pop() != '(') ||
                    (c == '}' && stack.pop() != '{') ||
                    (c == ']' && stack.pop() != '[')) {
                return false;
            }
        }
    return stack.isEmpty();
    }
}



// test cases
// Example 1:
// Input: s = "()"
// Output: true
// Example 2:
// Input: s = "()[]{}"
// Output: true
// Example 3:
// Input: s = "(]"
// Output: false
// Example 4:
// Input: s = "([])"
// Output: true

File: 24-01-2025/c++/ValidParentheses.cpp

class Solution {
public:
    bool isValid(string s) {
        stack<char>st;

        for(int i=0;i<s.length();i++){
            char c = s[i];
            if(c=='(' || c=='{' || c=='[')
            st.push(c);
            else{

                if(st.empty()) return false;
                char top = st.top();
                if((top == '(' && c==')') || (top=='{' && c=='}' ) || (top=='[' && c==']')){
                    st.pop();
                }
                else{
                    return false;
                }
   
            }
        }

        return st.empty();
    }
};


// Input: s = "()"

// Output: true

// Input: s = "()[]{}"

// Output: true

//TC : O(n)